ReactNativeClient/lib/services/searchengine/queryBuilder.ts

```diff
@@ -0,0 +1,361 @@
+	}
+
+	if (filters.has('-latitude')) {
+		makeLocationFilter('latitude', filters.get('-latitude'), queryParts, params, true, relation);
+	}
+
+	if (filters.has('longitude')) {
+		makeLocationFilter('longitude', filters.get('longitude'), queryParts, params, false, relation);
+	}
+
+	if (filters.has('-longitude')) {
+		makeLocationFilter('longitude', filters.get('-longitude'), queryParts, params, true, relation);
+	}
+
+	if (filters.has('altitude')) {
+		makeLocationFilter('altitude', filters.get('altitude'), queryParts, params, false, relation);
+	}
+
+	if (filters.has('-altitude')) {
+		makeLocationFilter('altitude', filters.get('-altitude'), queryParts, params, true, relation);
+	}
+
+	if (filters.has('title') || filters.has('body') || filters.has('text')) {
+		// there is something to fts search
+		queryParts.push(`${relation} notes_fts MATCH ?`);
+		let match: string[] = [];
+
+		if (filters.has('title')) {
+			match = match.concat(makeMatchQuery('title', filters));
+		}
+		if (filters.has('body')) {
+			match = match.concat(makeMatchQuery('body', filters));
+		}
+		if (filters.has('text')) {
+			match = match.concat(makeMatchQuery('text', filters));
+		}
+
+		if (relation === 'AND') { params.push(match.join(' ').trim()); }
+		if (relation === 'OR') { params.push(match.join(' OR ').trim()); }
+	}
+
+	if (filters.has('-text')) {
+		if (relation === 'AND') {
+			queryParts.push('AND ROWID NOT IN (SELECT ROWID FROM notes_fts WHERE notes_fts MATCH ?)');
+			params.push(filters.get('-text').join(' OR '));
+		}
+
+		if (relation === 'OR') {
+			filters.get('-text').forEach(value => {
+				queryParts.push('OR ROWID IN (select * from (select rowid from notes_fts except select rowid from notes_fts where notes_fts match ?))');
+				params.push(value);
+			});
+		}
+	}
+
+	if (filters.has('-title')) {
+		if (relation === 'AND') {
+			queryParts.push(`${relation} ROWID NOT IN (SELECT ROWID FROM notes_fts WHERE notes_fts.title MATCH ?)`);
+			params.push(filters.get('-title').join(' OR '));
+		}
+
+		if (relation === 'OR') {
+			filters.get('-title').forEach(value => {
+				queryParts.push('OR ROWID IN (select * from (select rowid from notes_fts except select rowid from notes_fts where notes_fts.title match ?))');
+				params.push(value);
+			});
+		}
+	}
+
+	if (filters.has('-body')) {
+		if (relation === 'AND') {
+			queryParts.push('AND ROWID NOT IN (SELECT ROWID FROM notes_fts WHERE notes_fts.body MATCH ?)');
+			params.push(filters.get('-body').join(' OR '));
+		}
+		if (relation === 'OR') {
+			filters.get('-body').forEach(value => {
+				queryParts.push('OR ROWID IN (select * from (select rowid from notes_fts except select rowid from notes_fts where notes_fts.body match ?))');
+				params.push(value);
+			});
+		}
+
+	}
+
+	if (withs.length > 0) {
+		query = ['WITH RECURSIVE' , withs.join(',') ,queryParts.join(' ')].join(' ');
+	} else {
+		query = queryParts.join(' ');
+	}
+
+	return { query, params };
```

Comment:
Likewise, the SQL should be over multiple lines. Also please either split the function into multiple sub-functions or avoid having duplicate code. Things like this for example:

```
if (value === '1') {
	queryParts.push(`${relation} ROWID IN (SELECT ROWID from notes_normalized where notes_normalized.is_todo = 1 AND notes_normalized.todo_completed != 0)`);
} else if (value === '0') {
	queryParts.push(`${relation} ROWID IN (SELECT ROWID from notes_normalized where notes_normalized.is_todo = 1 AND notes_normalized.todo_completed = 0)`);
```

There's only one character difference between the lines so surely that can be simplified, and there are several examples like this.

This pull request unfortunately is quite hard to review and that's partly because it's a tricky feature to implement, but you can also make it easier to understand (and maintain) by putting the SQL over several lines and by avoiding duplicate or unnecessary code.